### PR TITLE
Add setting for passing env var to terminal.

### LINKED ISSUE #14973: Set environment variable before starting a terminal
- VSCode Version: 1.7.1
- OS Version: Win 10 Pro

Add an option to set environment variables before starting the terminal.
Something like:
```json
"terminal.integrated.envVars.windows": ["DEBUG=1", "LANG=C"],
```
Or maybe you want to split variable and value something like:
```json
"terminal.integrated.envVars.windows": {
    "DEBUG": "1",
    "LANG": "C"
}
```

Obviously the same goes for Linux and Mac

### PR DESCRIPTION
**Pull request by ramya-rao-a**
Fixes #14973

### CODE REVIEW

File: src/vs/workbench/parts/terminal/common/terminal.ts [hunk 1]
```diff
@@ -67,6 +67,7 @@ export interface ITerminalConfiguration {
 	commandsToSkipShell: string[];
 	cwd: string;
 	confirmOnExit: boolean;
+	envVars: object;
```
Comment (Tyriar) [inline src/vs/workbench/parts/terminal/common/terminal.ts]: Let's call this `env`, that should be descriptive enough and it matches the `env` command.

Comment (Tyriar) [inline src/vs/workbench/parts/terminal/common/terminal.ts]: Also let's make the type more specific:

```ts
env: { [key: string]: string }
```

File: src/vs/workbench/parts/terminal/electron-browser/terminal.contribution.ts [hunk 2]
```diff
@@ -197,6 +197,21 @@ configurationRegistry.registerConfiguration({
 				DeleteWordLeftTerminalAction.ID,
 				DeleteWordRightTerminalAction.ID
 			].sort()
+		},
+		'terminal.integrated.envVars.osx': {
+			'description': nls.localize('terminal.integrated.envVars.osx', "Object with environment variables that will be passed to the terminal on OS X"),
```
Comment (Tyriar) [inline src/vs/workbench/parts/terminal/electron-browser/terminal.contribution.ts]: The description should be clear that this is adding these to the environment, not replacing the entire environment

File: src/vs/workbench/parts/terminal/electron-browser/terminalInstance.ts [hunk 3]
```diff
@@ -511,7 +511,9 @@ export class TerminalInstance implements ITerminalInstance {
 			this._configHelper.mergeDefaultShellPathAndArgs(shell);
 		}
 		this._initialCwd = this._getCwd(this._shellLaunchConfig, this._historyService.getLastActiveWorkspaceRoot());
-		const env = TerminalInstance.createTerminalEnv(process.env, shell, this._initialCwd, locale, this._cols, this._rows);
+		const platformKey = platform.isWindows ? 'windows' : platform.isMacintosh ? 'osx' : 'linux';
+		const envFromConfig = { ...process.env, ...this._configHelper.config.envVars[platformKey] };
```
Comment (Tyriar) [inline src/vs/workbench/parts/terminal/electron-browser/terminalInstance.ts]: Will this replace process.env entries with the envVars one when conflicts occur?

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/parts/terminal/electron-browser/terminalInstance.ts (hunk 3 above)
Comment (ramya-rao-a) [inline src/vs/workbench/parts/terminal/electron-browser/terminalInstance.ts]: yes

File: src/vs/workbench/parts/terminal/common/terminal.ts (hunk 1 above)
Comment (ramya-rao-a) [inline src/vs/workbench/parts/terminal/common/terminal.ts]: but if I do that I can't use my cool spread operator on it later :)

Comment (ramya-rao-a) [inline src/vs/workbench/parts/terminal/common/terminal.ts]: Done

File: src/vs/workbench/parts/terminal/electron-browser/terminal.contribution.ts (hunk 2 above)
Comment (ramya-rao-a) [inline src/vs/workbench/parts/terminal/electron-browser/terminal.contribution.ts]: Done

**Review by Tyriar — APPROVED** [review]
:ok_hand:

**Pull request merged**